File: src/main/java/szewek/fl/util/CapStorage.java
package szewek.fl.util;

import net.minecraft.nbt.NBTBase;
import net.minecraft.util.EnumFacing;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.util.INBTSerializable;

import javax.annotation.Nullable;

public final class CapStorage {
	private static final Capability.IStorage<?> EMPTY = new Capability.IStorage<Object>() {
		@Override
		public NBTBase writeNBT(Capability<Object> cap, Object t, EnumFacing side) {
			return null;
		}

		@Override
		public void readNBT(Capability<Object> cap, Object t, EnumFacing side, NBTBase nbt) {
		}
	};

	private static final Capability.IStorage<INBTSerializable<NBTBase>> NBT_STORAGE = new Capability.IStorage<INBTSerializable<NBTBase>>() {
		@Nullable
		@Override
		public NBTBase writeNBT(Capability<INBTSerializable<NBTBase>> cap, INBTSerializable<NBTBase> instance, EnumFacing side) {
			return instance.serializeNBT();
		}

		@Override
		public void readNBT(Capability<INBTSerializable<NBTBase>> cap, INBTSerializable<NBTBase> instance, EnumFacing side, NBTBase nbt) {
			instance.deserializeNBT(nbt);
		}
	};

	private static final Capability.IStorage<?> CUSTOM = new Capability.IStorage<Object>() {
		@Nullable
		@Override
		public NBTBase writeNBT(Capability<Object> capability, Object t, EnumFacing side) {
			return t instanceof INBTSerializable ? ((INBTSerializable) t).serializeNBT() : null;
		}

		@SuppressWarnings("unchecked")
		@Override
		public void readNBT(Capability<Object> capability, Object t, EnumFacing side, NBTBase nbt) {
			if (t instanceof INBTSerializable)
				((INBTSerializable<NBTBase>) t).deserializeNBT(nbt);
		}
	};

	@SuppressWarnings("unchecked")
	public static <T> Capability.IStorage<T> getEmpty() {
		return (Capability.IStorage<T>) EMPTY;
	}

	@SuppressWarnings("unchecked")
	public static <T extends INBTSerializable<NBTBase>> Capability.IStorage<T> getNBTStorage() {
		return (Capability.IStorage<T>) NBT_STORAGE;
	}

	@SuppressWarnings("unchecked")
	public static <T> Capability.IStorage<T> getCustom() {
		return (Capability.IStorage<T>) CUSTOM;
	}

	private CapStorage() {
	}
}

File: src/main/java/szewek/fl/util/CapStorage.kt
package szewek.fl.util

import net.minecraft.nbt.NBTBase
import net.minecraft.util.EnumFacing
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.util.INBTSerializable

@Suppress("UNCHECKED_CAST")
object CapStorage {
	private val EMPTY = object : Capability.IStorage<Any> {
		override fun writeNBT(cap: Capability<Any>, t: Any, side: EnumFacing): NBTBase? {
			return null
		}

		override fun readNBT(cap: Capability<Any>, t: Any, side: EnumFacing, nbt: NBTBase) {}
	}

	private val NBT_STORAGE = object : Capability.IStorage<INBTSerializable<NBTBase>> {
		override fun writeNBT(cap: Capability<INBTSerializable<NBTBase>>, instance: INBTSerializable<NBTBase>, side: EnumFacing): NBTBase? {
			return instance.serializeNBT()
		}

		override fun readNBT(cap: Capability<INBTSerializable<NBTBase>>, instance: INBTSerializable<NBTBase>, side: EnumFacing, nbt: NBTBase) {
			instance.deserializeNBT(nbt)
		}
	}

	private val CUSTOM = object : Capability.IStorage<Any> {
		override fun writeNBT(capability: Capability<Any>, t: Any, side: EnumFacing): NBTBase? {
			return (t as? INBTSerializable<*>)?.serializeNBT()
		}

		override fun readNBT(capability: Capability<Any>, t: Any, side: EnumFacing, nbt: NBTBase) {
			if (t is INBTSerializable<*>)
				(t as INBTSerializable<NBTBase>).deserializeNBT(nbt)
		}
	}

	@JvmStatic
	fun <T> getEmpty(): Capability.IStorage<T> {
		return EMPTY as Capability.IStorage<T>
	}

	@JvmStatic
	fun <T : INBTSerializable<NBTBase>> getNBTStorage(): Capability.IStorage<T> {
		return NBT_STORAGE as Capability.IStorage<T>
	}

	@JvmStatic
	fun <T> getCustom(): Capability.IStorage<T> {
		return CUSTOM as Capability.IStorage<T>
	}
}
